include/grpc/event_engine/memory_allocator.h

```diff
@@ -0,0 +1,242 @@
+    Reservation& operator=(Reservation&&) = default;
+    ~Reservation() {
+      if (allocator_ != nullptr) allocator_->Release(size_);
+    }
+
+   private:
+    friend class MemoryAllocator;
+    Reservation(std::shared_ptr<BasicMemoryAllocator> allocator, size_t size)
+        : allocator_(allocator), size_(size) {}
+
+    std::shared_ptr<BasicMemoryAllocator> allocator_ = nullptr;
+    size_t size_ = 0;
+  };
+
+  // Reserve bytes from the quota and automatically release them when
+  // Reservation is destroyed.
+  Reservation MakeReservation(MemoryRequest request) {
+    return Reservation(allocator_, Reserve(request));
+  }
+
+  // Allocate a new object of type T, with constructor arguments.
+  // The returned type is wrapped, and upon destruction the reserved memory
+  // will be released to the allocator automatically. As such, T must have a
+  // virtual destructor so we can insert the necessary hook.
+  template <typename T, typename... Args>
+  typename std::enable_if<std::has_virtual_destructor<T>::value, T*>::type New(
+      Args&&... args) {
+    // Wrap T such that when it's destroyed, we can release memory back to the
+    // allocator.
+    class Wrapper final : public T {
+     public:
+      explicit Wrapper(std::shared_ptr<BasicMemoryAllocator> allocator,
+                       Args&&... args)
+          : T(std::forward<Args>(args)...), allocator_(std::move(allocator)) {}
+      ~Wrapper() override { allocator_->Release(sizeof(*this)); }
+
+     private:
+      const std::shared_ptr<BasicMemoryAllocator> allocator_;
+    };
+    Reserve(sizeof(Wrapper));
+    return new Wrapper(allocator_, std::forward<Args>(args)...);
+  }
+
+  // Construct a unique ptr immediately.
+  template <typename T, typename... Args>
+  std::unique_ptr<T> MakeUnique(Args&&... args) {
+    return std::unique_ptr<T>(New<T>(std::forward<Args>(args)...));
+  }
+
+  // Allocate a slice, using MemoryRequest to size the number of returned bytes.
+  // For a variable length request, check the returned slice length to verify
+  // how much memory was allocated.
+  // Takes care of reserving memory for any relevant control structures also.
+  grpc_slice MakeSlice(MemoryRequest request);
+
+  // A C++ allocator for containers of T.
+  template <typename T>
+  class Container {
+   public:
+    using value_type = T;
+
+    // Construct the allocator: \a underlying_allocator is borrowed, and must
+    // outlive this object.
+    explicit Container(MemoryAllocator* underlying_allocator)
+        : underlying_allocator_(underlying_allocator) {}
+    template <typename U>
+    explicit Container(const Container<U>& other)
+        : underlying_allocator_(other.underlying_allocator()) {}
+
+    MemoryAllocator* underlying_allocator() const {
+      return underlying_allocator_;
+    }
+
+    T* allocate(size_t n) {
+      underlying_allocator_->Reserve(n * sizeof(T));
+      return static_cast<T*>(::operator new(n * sizeof(T)));
```

Comment:
Interesting. What happens if you remove the absolute qualification on `::operator new`? Concerns about ambiguity?

Reply:
Possibly... tbqh it's likely I cribbed this from whatever website I was reading to remember how to do this stuff.